lib/multi_aggregator/data_transfer.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+module MultiAggregator
+  class DataTransfer
+    attr_reader(
+      :logger
+    )
+
+    def initialize(
+      logger = MultiAggregator::Logger.new
+    )
+      @logger = logger
+    end
+
+    def call(query_spec, storage, providers = {})
+      query_spec.each do |provider_id, tables_spec|
+        transfer_tables(storage, providers[provider_id], provider_id, tables_spec)
+      end
+    end
+
+    private
+
+    def transfer_tables(storage, provider, db_name, tables_spec)
+      # TODO: move to processor
+      uid = generate_uid
+      storage.uid = uid
+
+      tables_spec.each do |table, columns_spec|
+        transfer_table(storage, provider, db_name, table, columns_spec)
+      end
```

Comment:
OK, good, and what do you think is the best solution? `run it in separated environments.` ?
If so could you provide more details here.

Reply:
What do you think about spawning new processes?
What do you think about detached queries?
Why do you think so? 
> but we probably will not get desired increasing performance because MRI etc

Reply:
>OK, good, and what do you think is the best solution?
>What do you think about spawning new processes?

I guess running in two environments is faster than running in one environment with two threads. Because MRI has GIL which is reducing speed for parallel execution.

>What do you think about detached queries?

Your question is not clear for me. Please specify which queries did you mean. In one DB? or different DB?

We can to face a problems in pool of DB/network bandwidth/RAM current process or DBMS/CPU/other specific DBMS nuances.

For example redis (without replications) works in one thread. So making parallel requests to redis is impractical.

Also we should care about DBMS becuase it can be using in other services in production. I mean that our parallel queries can slow down time execution of queries for other clients.

Do you need any additional explanations ?

Reply:
I added ability to copy async in one environment:
https://github.com/Hawk73/multi_aggregator_gem/pull/2
It has already merged.